tests/components/hyperion/test_light.py

```diff
@@ -1,82 +1,278 @@
     """Call a Hyperion entity callback that was registered with the client."""
     return client.set_callbacks.call_args[0][0][key](*args, **kwargs)
 
 
-async def setup_entity(hass, client=None):
+async def _setup_entity_yaml(hass, client=None):
     """Add a test Hyperion entity to hass."""
     client = client or create_mock_client()
     with patch("hyperion.client.HyperionClient", return_value=client):
-        assert await async_setup_component(
+        assert await setup.async_setup_component(
             hass,
-            DOMAIN,
+            LIGHT_DOMAIN,
             {
-                DOMAIN: {
+                LIGHT_DOMAIN: {
                     "platform": "hyperion",
-                    "name": TEST_NAME,
+                    "name": TEST_YAML_NAME,
                     "host": TEST_HOST,
-                    "port": const.DEFAULT_PORT,
+                    "port": TEST_PORT,
                     "priority": TEST_PRIORITY,
                 }
             },
         )
-    await hass.async_block_till_done()
+        await hass.async_block_till_done()
+
 
+async def test_setup_yaml_already_converted(hass):
+    """Test an already converted YAML style config."""
+    # This tests "Possibility 1" from async_setup_platform()
 
-async def test_setup_platform(hass):
-    """Test setting up the platform."""
+    # Add a pre-existing config entry.
+    add_test_config_entry(hass)
     client = create_mock_client()
-    await setup_entity(hass, client=client)
-    assert hass.states.get(TEST_ENTITY_ID) is not None
+    await _setup_entity_yaml(hass, client=client)
+
+    # Setup should be skipped for the YAML config as there is a pre-existing config
+    # entry.
+    assert hass.states.get(TEST_YAML_ENTITY_ID) is None
+
 
+async def test_setup_yaml_old_style_unique_id(hass):
+    """Test an already converted YAML style config."""
+    # This tests "Possibility 2" from async_setup_platform()
+    old_unique_id = f"{TEST_HOST}:{TEST_PORT}-0"
+
+    # Add a pre-existing registry entry.
+    registry = await async_get_registry(hass)
+    registry.async_get_or_create(
+        domain=LIGHT_DOMAIN,
+        platform=DOMAIN,
+        unique_id=old_unique_id,
+        suggested_object_id=TEST_YAML_NAME,
+    )
+
+    client = create_mock_client()
+    await _setup_entity_yaml(hass, client=client)
+
+    # The entity should have been created with the same entity_id.
+    assert hass.states.get(TEST_YAML_ENTITY_ID) is not None
+
+    # The unique_id should have been updated in the registry (rather than the one
+    # specified above).
+    assert registry.async_get(TEST_YAML_ENTITY_ID).unique_id == get_hyperion_unique_id(
+        TEST_SERVER_ID, 0
+    )
+    assert registry.async_get_entity_id(LIGHT_DOMAIN, DOMAIN, old_unique_id) is None
+
+    # There should be a config entry with the correct server unique_id.
+    entry_id = next(iter(hass.data[DOMAIN]))
+    assert client == hass.data[DOMAIN][entry_id][CONF_ROOT_CLIENT]
```

Comment:
Please don't assert any integration details.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations